quickwit/quickwit-indexing/src/throttle.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Weak};
+use std::time::{Duration, Instant};
+
+use quickwit_common::metrics::IntCounter;
+pub struct Throttle {
```

Comment:
We might not need a custom implementation: the `governor` rate limiter wraps an `AtomicU64`, so cloneable with an `Arc` and has sync (`check*`) and async `until*` APIs.

Reply:
One thing I am scared with governor is that 
a) the docs only talks about requests / operations etc. and does not talk about using it to throttle a throughput
b) the logic that throttles is a loop that have tasks wait for the capacity to be available. I am not fond of that:
I don't know what limits the number of times a thread will have to wake up, I am not sure if there can be some wait of time here.
c) I don't understand how it works so I am not able to reason with it:
For instance, if I have two writers, sharing a rate limiter, and being throttled. One is writing 1kB at a time, when the other is doing it 10kb at a time, what throughput do they achieve?

Reply:
created an issue on github: https://github.com/antifuchs/governor/issues/154

Reply:
created an issue on github: https://github.com/antifuchs/governor/issues/154

Reply:
https://github.com/tikv/async-speed-limit seems to do exactly what we need and more (can wrap async read/write directly).